Provider-agnostic message format (#270)
ueberdosis/tiptap-docs: src/content/content-ai/capabilities/agent/custom-llms/server-side-tools/openai-responses.mdx

title: Server-side tools (OpenAI Responses API)
meta:
  title: Server-side tools (OpenAI Responses API) | Tiptap Content AI

>   description: Learn how to call tools in the server using the OpenAI Responses API.

Reviewer: ```suggestion
  description: Learn how to call server-side tools using the OpenAI Responses API.
```

  category: Content AI
---

> This guide explains how to call tools in the server using the OpenAI Responses API. It presents an example of a tool that returns the weather in a given location.

Reviewer: ```suggestion
This guide explains how to call server-side tools using the OpenAI Responses API. It presents an example of a tool that returns the weather in a given location.
```

    continue
  }

  // Get weather from an external API
  const args = JSON.parse(toolCall.arguments)
  const result = getWeather(args)

  formatter.addChatMessage({
    type: 'toolCallResult',
    toolName: 'get_weather',
    toolCallId: toolCall.call_id,
    result: result.toString(),
  })
}

// Call the OpenAI Responses API again
const response = await openai.responses.create({
  model: 'gpt-4.1',
  input: [
    {
      role: 'developer',
      content: `
<Your system prompt>
${toolkit.getSystemPrompt()}
`,
    },
    ...formatter.format(),
  ],
  // Provide the tools that the AI model can call
  tools: toolkit.format(),
})
```

Finally, when there are no more server-side tool calls, use the `formatter` to convert the response to the format expected by the AI Agent extension.

```ts
const response = formatter.getResolverResponse()
```

> The `result` should be the response of the API endpoint, and the return value of the `resolver` function.

Reviewer: ```suggestion
The value returned from `formatter.getResolverResponse()` should be the response of your API endpoint and the return value of the resolver function.
```
